### PR TITLE
Basic localization support

### PR DESCRIPTION
**Pull request by woutersamaey**
@NicolasDorier please review the code in this PR as a basis for localization.

It is fully working with:
- Language switcher put in an easily visible place
- Locale stored in cookie as per .NET's best-practise, remembers for 1 year
- Fully compatible with .NET's IStringLocalizer (so no parallel solution - we can use native stuff)
- Data is stored in JSON files in a "Resources" folder, compatible with the Transiflex format
- Currently I have translated the main menu and left sidebar menus for the Dutch language so you can test and see content change
- English content is not affected as we discussed.
- All translation is done through 2 TagHelpers: 1 is for HTML content, and 1 is for translating HTML attributes like "title" and "alt".

Todo's are:
- Automatically detect the user's language based on the `Accept` request header as a last resort (if he doesn't have the cookie & no language stored in his user profile)
- Automatically append missing translations to a `en.json` file during development, so the file can be easily uploaded to Transiflex without needing to manually copy-paste keys. This is a pain.
- Automatically detect available languages (which JSON files are in the Resources dir?) - should be relatively easy
- More translations need to be added - also easy
- No translations with parameters encountered yet - not sure what to expect here
- No number formatting encountered yet - not sure what to expect here
- Text right-to-left support (like for Arabic) - appears not to be supported in Bootstrap HTML. Unofficial workarounds exist, but not sure which is best
- Sending localized emails in the user's language
- Translating form validation messages

Some things I'd like some help with are:
- Store the user's "preferred locale" in his profile. Would be essential for sending emails to the user in his language. Default to the language the user was using when creating his account or English for already existing users.
- ~~I'm not 100% sure the translations are kept in memory and the JSON file isn't re-read. Maybe you could help me find out? There is a flag to "reload on change" and that is set to "true", so maybe it's decent. I dunno.~~

I'm planning to address all of these TODOs, but would like to get your thought so far.

I don't expect you to approve the PR in this current state, though you probably can!! After all it is functional, just not 100% translated yet.

### CODE REVIEW

**Comment by NicolasDorier** [comment]
awesome work. reviewing

**Comment by NicolasDorier** [comment]
I think you should add a selenium test. This is quite easy to do and will properly test it works.

**Comment by NicolasDorier** [comment]
This is dope. So yes, only thing missing is the selenium test that will ensure your tagHelper works as desired.

**Comment by woutersamaey** [comment]
@NicolasDorier I have just committed the Selenium test but cannot run it myself. The selenium tests only run on Chrome 74, but I have 76. Not sure how to cope with this.
Can you check again please?

I also did a major performance improvement: Previously each translate call would re-create the translator and read the JSON file again and again, causing very slow page load. No longer an issue.

**Comment by NicolasDorier** [comment]
@woutersamaey that's strange, you should be able to run the tests. I will check.

**Comment by NicolasDorier** [comment]
You also need to fix the warnings by the way.

**Comment by NicolasDorier** [comment]
Check https://circleci.com/gh/btcpayserver/btcpayserver/2281?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link

**Comment by woutersamaey** [comment]
@NicolasDorier I'm seeing a lot of `Address not available` messages in CircleCI, but I don't think this has anything to do with this PR, am I right?

The warnings have been fixed, but the Selenium test still fails. Not sure why.

Since I can't debug (ChromeDriver version issue) it's hard to proceed from here. Any ideas to far?

**Comment by NicolasDorier** [comment]
Ignore `Address not available`.

**Comment by NicolasDorier** [comment]
```
Failed   BTCPayServer.Tests.ChromeTests.canLocalize
Error Message:
 Assert.Equal() Failure
          ↓ (pos 0)
Expected: Aanmelden
Actual:   login
          ↑ (pos 0)
Stack Trace:
   at BTCPayServer.Tests.ChromeTests.canLocalize() in /source/BTCPayServer.Tests/SeleniumTests.cs:line 393
```

**Comment by NicolasDorier** [comment]
if you are on windows, you can run selenium locally via docker by doing `cd .circleci && ./run-tests.sh "Selenium=Selenium"` but you can't debug :(

**Comment by woutersamaey** [comment]
I’m on Mac. 

Any reason why we can’t debug? Is it because of docker?

**Comment by NicolasDorier** [comment]
@woutersamaey can you share with me the exception you get when you try to run the selenium test?

**Comment by NicolasDorier** [comment]
@woutersamaey just for info: you run the docker-compose right?

**Comment by NicolasDorier** [comment]
working for me 76.0.3809.100

**Comment by woutersamaey** [comment]
@NicolasDorier this is the exception that I'm getting.
```
BTCPayServer.Tests.ChromeTests.canLocalize

System.InvalidOperationException : session not created: This version of ChromeDriver only supports Chrome version 74
  (Driver info: chromedriver=74.0.3729.6 (255758eccf3d244491b8a1317aa76e1ce10d57e9-refs/branch-heads/3729@{#29}),platform=Mac OS X 10.14.6 x86_64) (SessionNotCreated)
   at OpenQA.Selenium.Remote.RemoteWebDriver.UnpackAndThrowOnError(Response errorResponse)
   at OpenQA.Selenium.Remote.RemoteWebDriver.Execute(String driverCommandToExecute, Dictionary`2 parameters)
   at OpenQA.Selenium.Remote.RemoteWebDriver.StartSession(ICapabilities desiredCapabilities)
   at OpenQA.Selenium.Remote.RemoteWebDriver..ctor(ICommandExecutor commandExecutor, ICapabilities desiredCapabilities)
   at OpenQA.Selenium.Chrome.ChromeDriver..ctor(ChromeDriverService service, ChromeOptions options, TimeSpan commandTimeout)
   at BTCPayServer.Tests.SeleniumTester.Start() in /Users/wouter/Business/Workspace/btcpayserver/BTCPayServer.Tests/SeleniumTester.cs:line 42
   at BTCPayServer.Tests.ChromeTests.canLocalize() in /Users/wouter/Business/Workspace/btcpayserver/BTCPayServer.Tests/SeleniumTests.cs:line 379



8/20/19 7:23:10 AM +00:00 :Console:   Data Directory: /Users/wouter/Business/Workspace/btcpayserver/BTCPayServer.Tests/bin/Debug/netcoreapp2.1/canLocalize/pay
8/20/19 7:23:10 AM +00:00 :Console:   Configuration File: /Users/wouter/Business/Workspace/btcpayserver/BTCPayServer.Tests/bin/Debug/netcoreapp2.1/canLocalize/pay/RegTest/settings.config
8/20/19 7:23:12 AM +00:00 :Configuration:   Network: Regtest
8/20/19 7:23:12 AM +00:00 :Configuration:   Setting BTC.lightning is a deprecated format, it will work now, but please replace it for future versions with 'type=charge;server=http://127.0.0.1:54938/;api-token=foiewnccewuify'
8/20/19 7:23:12 AM +00:00 :Configuration:   Supported chains: BTC,LTC
8/20/19 7:23:12 AM +00:00 :Configuration:   Postgres DB used (User ID=postgres;Host=127.0.0.1;Port=39372;Database=btcpayserver)
8/20/19 7:23:15 AM +00:00 :Configuration:   Root Path: /
8/20/19 7:23:15 AM +00:00 :Configuration:   BTC: Explorer url is http://127.0.0.1:32838/
8/20/19 7:23:15 AM +00:00 :Configuration:   BTC: Cookie file is 0
8/20/19 7:23:15 AM +00:00 :Configuration:   BTC: Using cookie auth against NBXplorer, but 0 is not found
8/20/19 7:23:15 AM +00:00 :Configuration:   LTC: Explorer url is http://127.0.0.1:32838/
8/20/19 7:23:15 AM +00:00 :Configuration:   LTC: Cookie file is 0
8/20/19 7:23:15 AM +00:00 :Configuration:   LTC: Using cookie auth against NBXplorer, but 0 is not found
8/20/19 7:23:16 AM +00:00 :PayServer:   Starting listening NBXplorer (BTC)
8/20/19 7:23:16 AM +00:00 :PayServer:   Starting listening NBXplorer (LTC)
8/20/19 7:23:16 AM +00:00 :PayServer:   Start watching invoices
8/20/19 7:23:16 AM +00:00 :Events:   NBXplorer BTC: NotConnected => Ready
8/20/19 7:23:16 AM +00:00 :Events:   NBXplorer LTC: NotConnected => Ready
8/20/19 7:23:16 AM +00:00 :PayServer:   LTC: Checking if any pending invoice got paid while offline...
8/20/19 7:23:16 AM +00:00 :PayServer:   BTC: Checking if any pending invoice got paid while offline...
8/20/19 7:23:16 AM +00:00 :PayServer:   BTC: 0 payments happened while offline
8/20/19 7:23:16 AM +00:00 :PayServer:   LTC: 0 payments happened while offline
8/20/19 7:23:16 AM +00:00 :PayServer:   Connected to WebSocket of NBXplorer (LTC)
8/20/19 7:23:16 AM +00:00 :PayServer:   Connected to WebSocket of NBXplorer (BTC)
8/20/19 7:23:16 AM +00:00 :PayServer:   Starting payment request expiration watcher
8/20/19 7:23:16 AM +00:00 :Tests:   Listening on http://127.0.0.1:52945
8/20/19 7:23:16 AM +00:00 :Tests:   Server URI http://127.0.0.1:52945/
8/20/19 7:23:16 AM +00:00 :PayServer:   0 pending payment requests being checked since last run
8/20/19 7:23:19 AM +00:00 :PayServer:   Disconnected from WebSocket of NBXplorer (BTC)
8/20/19 7:23:19 AM +00:00 :PayServer:   Disconnected from WebSocket of NBXplorer (LTC)
8/20/19 7:23:19 AM +00:00 :PayServer:   Lightning listened stopped
8/20/19 7:23:19 AM +00:00 :PayServer:   Stop watching invoices
8/20/19 7:23:21 AM +00:00 :PayServer:   Error while contacting poloniex: The operation has timed out.
8/20/19 7:23:21 AM +00:00 :PayServer:   Error while contacting hitbtc: The operation has timed out.
8/20/19 7:23:23 AM +00:00 :PayServer:   Error while contacting bylls: The operation was canceled.
8/20/19 7:23:23 AM +00:00 :PayServer:   Error while contacting bitbank: The operation was canceled.
```

I am running like this (@rockstardev showed it to me this way during one of his onboarding sessions)
```
cd btcpayserver/BTCPayServer.Tests:
docker-compose up dev
```

These Docker containers have been running since last 3 weeks or so. Should I throw them away and re-create?

**Comment by NicolasDorier** [comment]
@woutersamaey can you rebase your PR on master? I updated the driver not long ago.

**Comment by woutersamaey** [comment]
The unit test is now working, but localization is failing for the Dutch language.
Since I can't debug, it appears there is an issue with the loading of the JSON resource file.

@NicolasDorier any idea if or how the loading of the JSON resource file could be different or fail when run as a unit test?

During a regular run or debug, everything works fine.
During the unit test, there appear to be no translations available (JSON Resource file issue).
The desired language is detected correctly and appearing in HTML. It just appears there are no translations loaded.

Debugging the unit test works (breakpoint in unit test code).
Debugging the app code does not work, there is some connection error.
This is what you meant with "Debugging isn't possible" right?

**Comment by woutersamaey** [comment]
@NicolasDorier the problem must be somewhere here during the unit test run.
Could it be that the resourcesPath differs when run as a unit test?
My guess is the JSON file can't be found.
```
var resourceFile = $"{culture}.json";
var resourcesPath = Path.Combine(PathHelpers.GetApplicationRoot(), GetResourcePath(assembly));
_searchedLocation = Path.Combine(_resourcesPath, resourceFile);

if (File.Exists(_searchedLocation)){
    // Load translations from JSON => Does not happen when run as unit test
}
```

**Comment by woutersamaey** [comment]
@NicolasDorier Any ideas? I have no idea how to handle this.

**Comment by NicolasDorier** [comment]
@woutersamaey yeah I can probably fix that give me a sec.

**Comment by NicolasDorier** [comment]
@woutersamaey I fixed your tests, but you need to give me right to your branch. You should see it on the right side of this page a checkbox to allow me to edit.

**Comment by NicolasDorier** [comment]
You will need to find a better place where to put the dropbox after I add my commit. This break the design.

**Comment by woutersamaey** [comment]
@NicolasDorier the checkbox is already checked?
image

Can you show me how your layout broke? On my end it looks quite OK and is in line with Bootstrap HTML rules.

**Comment by NicolasDorier** [comment]
@woutersamaey you can now pull my commit, the tests should pass.

image

I am talking about that, this is horrible.

**Comment by woutersamaey** [comment]
@NicolasDorier Where do you suggest I put it?

I can put it in the footer if you prefer that, but in that case I'll make sure to add automatic language detection based on the `Accept` header since the switcher won't be as visible as below the fold for most users.

**Comment by woutersamaey** [comment]
Pulled the code and it is working! Tnx for your help!

**Comment by NicolasDorier** [comment]
@woutersamaey idea:

image

**Comment by woutersamaey** [comment]
@NicolasDorier I like the minimalistic presentation, but don't think the globe is well recognized for translation.

Why not keep the dropdown but with:
- The language code EN / NL as labels
- Use CSS to make the `<select>` blend with the background so you can only see the 2 letter label (no white background, no border, uppercase)

I prefer to keep the dropdown as it triggers a form HTTP POST + it is convenient on mobile devices thanks to native input controls.

Having a traditional menu dropdown or a pop-up modal would be a problem once we get like 40 languages.

What do you think?

**Comment by NicolasDorier** [comment]
I am not good in html/css but I think that you can make the globe (or any other logo you like in font awesome) show a dropdown when clicked without javascript.

> Having a traditional menu dropdown or a pop-up modal would be a problem once we get like 40 languages.

We'll see what do to when we get to this point. Though I don't think this is problematic.

> If you think changing the CSS is sufficient, I would prefer to keep the languages names in full form and not 2 letter codes. Maybe add a separator on the left, that could also help set it aside from the main menu.

Whatever, but it should not take too much screen space as it is rare people use it once they have set their location. Another idea is to put this in the `User profile`.

**Comment by NicolasDorier** [comment]
You can browse icons on https://fontawesome.com/icons

**Comment by NicolasDorier** [comment]
Maybe in the footer it can also work.

**Comment by woutersamaey** [comment]
@NicolasDorier I'll do a combination of what you suggested.

Full dropdown in footer (best to keep this for older people) + globe icon on top.

You're right about the dropdown without JavaScript and I am familiar with FontAwesome. This is kinda my thing 👍

File: BTCPayServer.Tests/docker-compose.yml [hunk 1]
```diff
@@ -119,7 +119,7 @@ services:
       - "bitcoin_datadir:/data"
 
   customer_lightningd:
-    image: btcpayserver/lightning:v0.7.1-dev
+    image: btcpayserver/lightning:v0.7.2-dev
```
Comment (Kukks) [inline BTCPayServer.Tests/docker-compose.yml]: I think you need to rebase

**Comment by NicolasDorier** [comment]
@woutersamaey please rebase. We will use the current dropbox only in regtest (hidding it on mainnet) until you can replace it with something better.

**Comment by woutersamaey** [comment]
Will do. Please give me 1-2 days. My holiday is over and I'm back at work so busier again :(

**Comment by woutersamaey** [comment]
Unfortunately I'm not finding the time to continue work on this. I got pretty far, but lacking the time to finish.
If someone could help out to push this though, I'd be very grateful.
I'm free to do a call if it helps.

**Comment by Kukks** [comment]
Will take it upon myself to finish this as soon as I'm free

**Comment by rockstardev** [comment]
I'll take this over... will close it temporarily as we clean up our PRs (gotta reach 0 @NicolasDorier). Will rebase and merge existing effort as soon as I have time.

**Comment by Kukks** [comment]
Reopening to reconsider

**Comment by Spunsd** [comment]
I minted my BTC already, can it now?

**Pull request closed without merging**